File: packages/scoring-rules/classify_archetype.py
"""
Archetype classifier — heuristic routing of job descriptions to role archetypes.

Reads packages/scoring-rules/archetypes.yaml and returns the best-matching
archetype along with a confidence score. Used by job_score.py to apply
archetype-specific strictness to the scoring bands.

The TypeScript mirror lives at apps/web/src/scoring/classify-archetype.ts.
Keep the two in sync.
"""
from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any

try:
    import yaml
except ImportError:
    yaml = None

DEFAULT_CONFIG = Path(__file__).resolve().parent / "archetypes.yaml"


@dataclass
class Classification:
    archetype: str
    label: str
    confidence: float
    strictness: float
    matched_signals: list[str]

    def to_dict(self) -> dict[str, Any]:
        return {
            "archetype": self.archetype,
            "label": self.label,
            "confidence": round(self.confidence, 3),
            "strictness": self.strictness,
            "matched_signals": self.matched_signals,
        }


def load_config(path: Path | None = None) -> dict:
    if yaml is None:
        raise RuntimeError("PyYAML required for archetype classifier")
    cfg_path = path or DEFAULT_CONFIG
    if not cfg_path.exists():
        raise FileNotFoundError(f"archetypes config missing: {cfg_path}")
    with cfg_path.open("r", encoding="utf-8") as f:
        return yaml.safe_load(f) or {}


def _score_archetype(name: str, rules: dict, title: str, jd: str) -> tuple[float, list[str]]:
    """Return (score 0-1, matched signal names) for a single archetype."""
    signals: list[str] = []
    hits = 0.0
    total_weight = 0.0

    # Title patterns — weight 0.5 (most specific)
    title_patterns = rules.get("title_patterns", []) or []
    if title_patterns:
        total_weight += 0.5
        for pat in title_patterns:
            if re.search(pat, title):
                hits += 0.5
                signals.append(f"title:{pat[:40]}")
                break

    # Required keywords — weight 0.4
    required = rules.get("required_keywords", []) or []
    if required:
        total_weight += 0.4
        for pat in required:
            if re.search(pat, jd):
                hits += 0.4
                signals.append(f"keyword:{pat[:40]}")
                break

    # Disqualifiers — any hit zeroes the score
    disq = rules.get("disqualifiers", []) or []
    for pat in disq:
        if re.search(pat, jd) or re.search(pat, title):
            return (0.0, [f"disqualified:{pat[:40]}"])

    if total_weight == 0:
        return (0.0, signals)
    return (hits / total_weight, signals)


def classify_archetype(
    title: str,
    jd_text: str = "",
    hints: list[str] | None = None,
    config: dict | None = None,
) -> Classification:
    """
    Classify a job description into an archetype.

    Args:
        title: Job title (most important signal)
        jd_text: Full job description text
        hints: Optional pre-labels from portal_targets.yaml (e.g. ["ai-applied"])
        config: Pre-loaded config (for testing); falls back to archetypes.yaml
    """
    cfg = config if config is not None else load_config()
    min_conf = float(cfg.get("min_confidence", 0.6))
    archetypes = cfg.get("archetypes", {})
    fallback = cfg.get("fallback", {"label": "General", "strictness": 1.0})

    title = title or ""
    jd = jd_text or ""

    # If hints are provided and score >= threshold for that specific hint, use it.
    candidates: list[tuple[str, float, list[str]]] = []
    hint_set = set(hints or [])

    for name, rules in archetypes.items():
        score, signals = _score_archetype(name, rules, title, jd)
        if name in hint_set and score > 0:
            # Pre-label from portal_targets adds 0.15 boost, capped at 1.0
            score = min(1.0, score + 0.15)
            signals = signals + [f"hint:{name}"]
        candidates.append((name, score, signals))

    # Sort by (score DESC, priority DESC) — niche archetypes beat generic ones
    # on ties so "Founding AI Engineer" routes to founder-minded-ic and a
    # healthcare CTO routes to healthcare-ops, not generic leadership.
    def _priority(name: str) -> int:
        return int(archetypes.get(name, {}).get("priority", 0))

    candidates.sort(key=lambda c: (c[1], _priority(c[0])), reverse=True)
    best_name, best_score, best_signals = candidates[0]

    if best_score < min_conf:
        return Classification(
            archetype="general",
            label=fallback.get("label", "General"),
            confidence=best_score,
            strictness=float(fallback.get("strictness", 1.0)),
            matched_signals=best_signals,
        )

    best_rules = archetypes[best_name]
    return Classification(
        archetype=best_name,
        label=best_rules.get("label", best_name),
        confidence=best_score,
        strictness=float(best_rules.get("strictness", 1.0)),
        matched_signals=best_signals,
    )


def apply_archetype_to_bands(match_pct: float, gap_count: int, strictness: float) -> str:
    """
    Given a raw match_pct (0-1) and gap count, return the score band,
    adjusted by the archetype's strictness multiplier.

    strictness 1.0 = base bands (0.8 / 0.6 / 0.4)
    strictness 1.1 = stricter bands (0.88 / 0.66 / 0.44)
    strictness 0.9 = more lenient (0.72 / 0.54 / 0.36)
    """
    # Round to 4 decimal places to avoid float-precision edge cases
    # (e.g., 0.8 * 1.1 = 0.8800000000000001, blocking match_pct = 0.88).
    strong_t = round(0.8 * strictness, 4)
    good_t = round(0.6 * strictness, 4)
    stretch_t = round(0.4 * strictness, 4)

    if match_pct >= strong_t and gap_count == 0:
        return "strong"
    if match_pct >= good_t and gap_count <= 1:
        return "good"
    if match_pct >= stretch_t and gap_count <= 2:
        return "stretch"
    return "long_shot"

File: apps/web/src/scoring/classify-archetype.ts
/**
 * Archetype classifier — TypeScript mirror of packages/scoring-rules/classify_archetype.py
 *
 * Heuristic routing of job descriptions to role archetypes. Rules live in
 * packages/scoring-rules/archetypes.yaml. When you change one, change both.
 */

export type ArchetypeName =
  | "engineering-leadership"
  | "ai-applied"
  | "data-analytics"
  | "platform-sre"
  | "founder-minded-ic"
  | "security"
  | "healthcare-ops"
  | "general";

export interface ArchetypeRules {
  label: string;
  description?: string;
  strictness: number;
  priority?: number;
  title_patterns: string[];
  required_keywords: string[];
  disqualifiers: string[];
}

export interface ArchetypeConfig {
  version: number;
  min_confidence: number;
  archetypes: Record<Exclude<ArchetypeName, "general">, ArchetypeRules>;
  fallback: { label: string; strictness: number };
}

export interface Classification {
  archetype: ArchetypeName;
  label: string;
  confidence: number;
  strictness: number;
  matched_signals: string[];
}

/**
 * Hardcoded mirror of archetypes.yaml. A build step (future) will generate this
 * from the YAML. Today, keep it in manual sync — check with doctor.py's
 * scoring-sync check and tests/scoring/classify-archetype.test.ts.
 */
export const ARCHETYPE_CONFIG: ArchetypeConfig = {
  version: 1,
  min_confidence: 0.6,
  archetypes: {
    "engineering-leadership": {
      label: "Engineering Leadership",
      strictness: 1.0,
      title_patterns: [
        "^(?:vp|svp|head of|director)\\b",
        "\\bdirector[,\\s]+(?:software|engineering|platform|product)\\b",
        "\\b(?:engineering manager|em|sr\\.?\\s+manager)\\b",
        "\\bhead of (?:engineering|platform|product engineering)\\b",
        "\\b(?:cto|chief technology officer)\\b",
        "\\bsr\\.?\\s+director\\b",
      ],
      required_keywords: [
        "\\b(?:team|org|people|direct reports|managers?|leadership)\\b",
      ],
      disqualifiers: ["\\bindividual contributor only\\b"],
    },
    "ai-applied": {
      label: "Applied AI / ML Engineering",
      strictness: 1.0,
      title_patterns: [
        "\\b(?:ai engineer|ml engineer|applied (?:ai|ml)|llm engineer)\\b",
        "\\b(?:head of|director of|principal|staff) (?:ai|ml|applied ai)\\b",
        "\\bai (?:platform|product|infrastructure)\\b",
        "\\bagent(ic)? (?:engineer|developer)\\b",
      ],
      required_keywords: [
        "\\b(?:llm|gpt|claude|langchain|openai|anthropic|rag|retrieval|embeddings?|vector|fine[- ]?tun|prompt)\\b",
      ],
      disqualifiers: [
        "\\bresearch scientist\\b",
        "\\bphd\\s+(?:required|mandatory)\\b",
      ],
    },
    "data-analytics": {
      label: "Data / Analytics Engineering",
      strictness: 1.0,
      title_patterns: [
        "\\b(?:senior|staff|principal|lead) data engineer\\b",
        "\\banalytics engineer\\b",
        "\\bhead of (?:data|analytics|bi)\\b",
        "\\bdata platform (?:engineer|lead)\\b",
      ],
      required_keywords: [
        "\\b(?:dbt|snowflake|bigquery|redshift|airflow|spark|warehouse|etl|elt)\\b",
      ],
      disqualifiers: [],
    },
    "platform-sre": {
      label: "Platform / SRE / Infrastructure",
      strictness: 1.0,
      title_patterns: [
        "\\b(?:senior|staff|principal) (?:platform|infrastructure|sre|site reliability|devops) engineer\\b",
        "\\bhead of (?:platform|infrastructure|sre|devex)\\b",
        "\\bplatform (?:engineer|lead)\\b",
      ],
      required_keywords: [
        "\\b(?:kubernetes|terraform|aws|gcp|observability|reliability|slo|on[- ]call)\\b",
      ],
      disqualifiers: [],
    },
    "founder-minded-ic": {
      label: "Founder-minded IC",
      strictness: 0.9,
      priority: 10,
      title_patterns: [
        "\\bfounding (?:engineer|ai engineer|platform engineer|swe)\\b",
        "\\b(?:first|early) engineer\\b",
        "\\b0[- ]to[- ]1\\b",
      ],
      required_keywords: [
        "\\b(?:founding|early stage|seed|series a|0[- ]to[- ]1|first\\s+hire|small team)\\b",
      ],
      disqualifiers: [],
    },
    security: {
      label: "Security Engineering",
      strictness: 1.0,
      title_patterns: [
        "\\b(?:senior|staff|principal) (?:security|appsec) engineer\\b",
        "\\bhead of (?:security|product security|appsec)\\b",
        "\\bciso\\b",
      ],
      required_keywords: [
        "\\b(?:owasp|threat model|siem|pentest|vulnerability|ciso|soc\\s?2|iso\\s?27001)\\b",
      ],
      disqualifiers: [],
    },
    "healthcare-ops": {
      label: "Healthcare / Regulated Ops",
      strictness: 1.0,
      priority: 10,
      title_patterns: [
        "\\b(?:cto|vp engineering|head of engineering)\\b",
      ],
      required_keywords: [
        "\\b(?:hipaa|phi|fhir|hl7|ehr|emr|hitrust|fda|cms)\\b",
      ],
      disqualifiers: [],
    },
  },
  fallback: { label: "General", strictness: 1.0 },
};

function scoreArchetype(
  rules: ArchetypeRules,
  title: string,
  jd: string
): { score: number; signals: string[] } {
  const signals: string[] = [];
  let hits = 0;
  let totalWeight = 0;

  if (rules.title_patterns.length) {
    totalWeight += 0.5;
    for (const p of rules.title_patterns) {
      if (new RegExp(p, "i").test(title)) {
        hits += 0.5;
        signals.push(`title:${p.slice(0, 40)}`);
        break;
      }
    }
  }

  if (rules.required_keywords.length) {
    totalWeight += 0.4;
    for (const p of rules.required_keywords) {
      if (new RegExp(p, "i").test(jd)) {
        hits += 0.4;
        signals.push(`keyword:${p.slice(0, 40)}`);
        break;
      }
    }
  }

  for (const p of rules.disqualifiers) {
    const rx = new RegExp(p, "i");
    if (rx.test(jd) || rx.test(title)) {
      return { score: 0, signals: [`disqualified:${p.slice(0, 40)}`] };
    }
  }

  if (totalWeight === 0) return { score: 0, signals };
  return { score: hits / totalWeight, signals };
}

export function classifyArchetype(
  title: string,
  jdText = "",
  hints: string[] = [],
  config: ArchetypeConfig = ARCHETYPE_CONFIG
): Classification {
  const safeTitle = title || "";
  const safeJd = jdText || "";
  const hintSet = new Set(hints);
  const candidates: Array<{
    name: ArchetypeName;
    score: number;
    signals: string[];
  }> = [];

  for (const [name, rules] of Object.entries(config.archetypes)) {
    const result = scoreArchetype(rules, safeTitle, safeJd);
    if (hintSet.has(name) && result.score > 0) {
      result.score = Math.min(1, result.score + 0.15);
      result.signals.push(`hint:${name}`);
    }
    candidates.push({
      name: name as ArchetypeName,
      score: result.score,
      signals: result.signals,
    });
  }

  // Sort by (score DESC, priority DESC) — niche archetypes beat generic on ties.
  const priorityOf = (name: ArchetypeName): number => {
    const rules = (config.archetypes as Record<string, ArchetypeRules | undefined>)[name];
    return rules?.priority ?? 0;
  };
  candidates.sort((a, b) => {
    if (b.score !== a.score) return b.score - a.score;
    return priorityOf(b.name) - priorityOf(a.name);
  });
  const best = candidates[0];

  if (best.score < config.min_confidence) {
    return {
      archetype: "general",
      label: config.fallback.label,
      confidence: Number(best.score.toFixed(3)),
      strictness: config.fallback.strictness,
      matched_signals: best.signals,
    };
  }

  const rules = config.archetypes[best.name as Exclude<ArchetypeName, "general">];
  return {
    archetype: best.name,
    label: rules.label,
    confidence: Number(best.score.toFixed(3)),
    strictness: rules.strictness,
    matched_signals: best.signals,
  };
}

/**
 * Given raw match_pct (0-1) and gap count, return the score band adjusted by
 * archetype strictness. Mirrors apply_archetype_to_bands() in Python.
 */
export function applyArchetypeToBands(
  matchPct: number,
  gapCount: number,
  strictness: number
): "strong" | "good" | "stretch" | "long_shot" {
  // Round to 4 decimals to dodge float-precision (0.8 * 1.1 = 0.8800000000000001).
  const strongT = Number((0.8 * strictness).toFixed(4));
  const goodT = Number((0.6 * strictness).toFixed(4));
  const stretchT = Number((0.4 * strictness).toFixed(4));

  if (matchPct >= strongT && gapCount === 0) return "strong";
  if (matchPct >= goodT && gapCount <= 1) return "good";
  if (matchPct >= stretchT && gapCount <= 2) return "stretch";
  return "long_shot";
}
